gauth/gauth.go

```diff
@@ -24,15 +25,37 @@ func IndexNow() (int64, int) {
 	return time / 30, int(time % 30)
 }
 
+func newConfig(digits int, hashName string) (otp.Config, error) {
```

Comment:
For better separation of concerns, I recommend not combining hash algorithm selection with construction of the config. As an alternative, consider maybe:

```go
// pickAlgorithm returns a constructor for the named hash function, or
// nil if the name is not a supported algorithm.
func pickAlgorithm(name string) func() hash.Hash {
   switch name {
   case "", "SHA1":
      return sha1.New
   case "SHA256":
      return sha256.New
   case "SHA512":
      return sha512.New
   default:
      return nil
   }
}
```

Then below:

```go
alg := pickAlgorithm(u.Algorithm)
if alg == nil {
   return "", "", "", fmt.Errorf("unsupported algorithm: %q", u.Algorithm)
}
cfg := otp.Config{Hash: alg, Digits: u.Digits}
```

The effect is the same, but that focuses the helper on one task rather than two.

Reply:
i like this suggestion. But i will do `func pickAlgorithm(name string) func() hash.Hash, error` instead.